### ISSUE TITLE
dual stack socket broken on macOS

### ISSUE THREAD

**Issue by XAMPPRocky** [googleforgames/quilkin#816]
On macOS (both Sonoma and whatever was before, and probably all other versions), using IPv6 has just been broken. Like the same code will work perfectly fine on Linux, but on macOS it will just never work.

To make development on macOS easier, we should disable dual stack, and have it be a Linux only feature, since it's the only one with a working stack as far as I can tell. This isn't just a quilkin issue, I've had trouble just trying `ping6` over IPv6 on macOS to google.com.

For whatever reason, it just never works, it's not an error, it just gets stuck.

**Comment by markmandel** [comment]
You probably have checked this, but is ipv6 enabled in your mac?
https://support.apple.com/en-au/guide/mac-help/mchlp2499/mac

It seems very strange to me that you can't `ping6` anything.

**Comment by XAMPPRocky** [comment]
It's set automatically but I don't see a prefix, so maybe it's just disabled on my mac 🤔

Wonder if we could query that and present a warning.

**Comment by markmandel** [comment]
Also curious, if you run `ifconfig` (I assume you can do that on a mac), do you have a local ipv6 address on your network, or just ipv4? (I have both on everything)

I do wonder how we can check these things 🤔 who is to know if the issue is the local machine, the router or something else that is blocking ipv6 traffic.

I checked with a buddy with a Mac, and he also got no response on `ping6 google.com`, but he has specifically disable ipv6 on his router.

That being said, since we have the Socket wrapper, I don't think it would be super hard to have a feature in Quilkin that says "only use ipv4", (`network=ipv4|ipv6` ?) which can be used at `cargo test` time if needed (such as in your example), or if someone needed to lock things down to only ipv4 they could do so when building a binary (although not sure if we would ship one?).

**Comment by markmandel** [comment]
So just for fun, I polled a bunch of teammates with macs.

Some got:

```
➜  ~ ping6 google.com
PING6(56=40+8+8 bytes) 2621:0:1002:a:f991:8ec5:5812:9b3d --> 2607:f8b0:4005:813::200e
16 bytes from 2607:f8b0:4005:813::200e, icmp_seq=0 hlim=120 time=3.931 ms
16 bytes from 2607:f8b0:4005:813::200e, icmp_seq=1 hlim=120 time=4.592 ms
16 bytes from 2607:f8b0:4005:813::200e, icmp_seq=2 hlim=120 time=3.403 ms
16 bytes from 2607:f8b0:4005:813::200e, icmp_seq=3 hlim=120 time=3.908 ms
16 bytes from 2607:f8b0:4005:813::200e, icmp_seq=4 hlim=120 time=4.169 ms
16 bytes from 2607:f8b0:4005:813::200e, icmp_seq=5 hlim=120 time=3.866 ms
^C
--- google.com ping6 statistics ---
6 packets transmitted, 6 packets received, 0.0% packet loss
round-trip min/avg/max/std-dev = 3.403/3.978/4.592/0.357 ms
```

Some got

```
$ ping6 google.com
ping6: UDP connect: No route to host
```

and some got 

```
$ ping6 google.com
PING6(56=40+8+8 bytes) fdf2:311e:6600:1:cc2:388f:3341:87a7 --> 2607:f8b0:400a:807::200e
```

and then nothing happened (ipv6 disabled at router).

So there seems to be a wide variety of possibilities here. Having an ipv4 only feature seems like the best way forward I think, since it's likely not just a mac issue.

**Comment by markmandel** [comment]
Doing some more research (and thinking some more) - https://crates.io/crates/local-ip-address might also be a good approach.

A command line arguement/env var is probably better than a feature flag (across the binary and testing). We could probably also use https://crates.io/crates/local-ip-address to check if the local machine has an ipv6 address as an auto detect (if it has one, assume ipv6 connectivity, if not, use ipv4).

This will disable us from doing some tests on an IPv4 only network, but it at least unblocks development, and the rest can be handled by CI.

**Comment by XAMPPRocky** [comment]
> A command line arguement/env var is probably better than a feature flag (across the binary and testing)

Definitely agree, if it's a feature flag, that means you have to make a custom build, or we ship two versions, and even then you'd have to know that this is an issue to find that solution. Having the program itself try to autodetect, warn when it can't find either kind of address, and provide a flag to force it to use one or the other (the default would be both).

> This will disable us from doing some tests on an IPv4 only network, but it at least unblocks development, and the rest can be handled by CI.

Couldn't we have tests for IPv4 only? Like we couldn't run every test but we could at least have a set of tests that need to work across, IPv4, IPv6, and dual stack. Btw I'm not totally blocked as I got it working on my WSL which seems to actually work with IPv6.

**Comment by markmandel** [comment]
> Couldn't we have tests for IPv4 only? Like we couldn't run every test but we could at least have a set of tests that need to work across, IPv4, IPv6, and dual stack.

Oh yeah, 100% agree. The tests we can't run are the tests that specifically test ipv4+ipv6 dual stack - and that's totally fine. I think we're saying the same thing here.

Weird that WSL is working for you, but a Mac network doesn't. This was what also led me to a command line/env var - since if we ever want to actually have this run client side, we need this as an option with some autodetect options as well.

**Comment by XAMPPRocky** [comment]
> Weird that WSL is working for you, but a Mac network doesn't.

Yeah and the underlying windows machine also does not work with IPv6 😄

**Comment by markmandel** [comment]
> Yeah and the underlying windows machine also does not work with IPv6 😄

![What the?](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExZjFnZHBhaGFldGJxa3pnNjRoOWE0NmZhc3pqZHV4MjExNjhjMG1teiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/3ohhweXI8WsSXXdWzS/giphy-downsized-large.gif)
